### PR TITLE
fix: relax sandboxing check for dir targets

### PR DESCRIPTION
**Pull request by rgrinberg**
when a rule doesn't need sandboxing, it should still be able to produce
directory targets.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

ps-id: ca89fead-e8c3-41f1-a861-3916267fa439

### CODE REVIEW

File: src/dune_engine/build_system.ml [hunk 1]
```diff
@@ -476,10 +482,12 @@ end = struct
           in
           let produced_targets =
             match sandbox with
-            | None ->
+            | None -> (
               (* Directory targets are not allowed for non-sandboxed actions, so
                  the call below should not raise. *)
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:487]: That's a stale comment, right? The above change allows directory targets for some non-sandboxed actions.

File: src/dune_engine/build_system.ml [hunk 2]
```diff
@@ -434,7 +434,13 @@ end = struct
            without sandboxing. We just need to make sure we clean up all stale
            directory targets before running the rule and then we can discover
            all created files right in the build directory. *)
-        if not (Path.Build.Set.is_empty targets.dirs) then
+        if
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:437]: As the above comment says, we might need to do some preparation before running a non-sandboxed action that produces directory targets. What if those directories already contain some files, for example?

**Review by snowleopard — COMMENTED** [review]
Thanks for factoring out this change! It's much easier for me to think about it in isolation.

Could you add a test to make sure this relaxation actually works?

**Comment by rgrinberg** [comment]
Thanks for reviewing. I've also made an alternative to this PR: https://github.com/ocaml/dune/pull/6056

It's a bit more complex, but also much more useful in general. In that PR I've allowed all non sandboxed actions to produce directory targets.

**Comment by snowleopard** [comment]
@rgrinberg I think I prefer the other, more general PR. Shall we close this one then?

**Comment by rgrinberg** [comment]
Yeah, I prefer the other PR as well.

**Pull request closed without merging**